### ISSUE TITLE
`@override` assignability not enforced for `__init__`

### ISSUE THREAD

**Issue by randolf-scholz** [astral-sh/ty#3083]
### Describe the Bug

[Playground](https://play.ty.dev/cb81c6b3-b847-4a77-8b86-608a24188252)

```python
from typing import override

class Foo:
    def __init__(self, x: int) -> None: ...
    
    
class Bar(Foo):
    @override
    def __init__(self, y: int) -> None: ...  # no error
```

Typically type checkers do no check that subclass `__init__` is assignable to superclass `__init__`. However, this check should happen when the method is explicitly annotated with `@override` because according to the [typing spec](https://typing.python.org/en/latest/spec/class-compat.html#override):

> When type checkers encounter a method decorated with @typing.override they should treat it as a type error unless that method is overriding a method or attribute in some ancestor class, **and the type of the overriding method is [assignable](https://typing.python.org/en/latest/spec/glossary.html#term-assignable) to the type of the overridden method.**

**Comment by sharkdp** [comment]
Thank you for reporting this. It doesn't look like any other type checker behaves like this (i.e. emits an error on this snippet)?

X-ref: https://github.com/python/typing/issues/2222

**Comment by charliermarsh** [comment]
Fixed by https://github.com/astral-sh/ruff/pull/28115.
